### ISSUE TITLE
Ability to mark/hide declined events

### ISSUE THREAD

**Issue by hobarrera** [pimutils/khal#580]
I currently have some work events which I've declined, and are indistinguishable from accepted events.

I'd like to somehow mark (an icon similar to recurring events), hide, stike-through (this is my first choice, but very few terminals support it, including my own) or dim (eg: slightly darker colour) events which I've declined.

As far as I understand, this also requires khal to know my email[s] (because that's how attendees are identified).

IMHO, dimming or some special character to recognize these are the best choices.

**Comment by geier** [comment]
Good idea, once khal know about your emails, it can also answer(RSVP) and send them (invite).

I have actually yesterday implemented something similar for `STATUS=CANCELLED` events, which for now are displayed with a big, fat CANCELLED in front, which is indeed very ugly.



Your suggestions

* strikethrough: I like this best, and my *terminal* supports this, but tmux which I often use to run khal etc in does not (and it looks like [they are not interested to](https://github.com/tmux/tmux/issues/612)), so this is probably out (to test run ``echo -e "\e[9mstrikethrough\e[0m" ``)

* dimmed: I find dimmed (dark gray for dark background, light gray for light background) problematic, because it often leads to me not being able to read that part, than I need more attention to actually read it (which kind of defeats the purpose). But perhaps I would just get used to ignoring unreadable stuff in khal.

* symbol: 
  1. do you have a suggestion for a symbol?
  2. is that noticable enough? (also for cancelled events)

**Comment by geier** [comment]
An alternative to ANSI codes for strikethrough would be [unicode strikethrough](https://en.wikipedia.org/wiki/Strikethrough#Combining_characters).

Seems to work for me:

```python
In [1]: STRIKE = '\u0336'

In [2]: def strikethrough(string):
   ...:     new = list()
   ...:     for char in string:
   ...:         new.append(STRIKE)
   ...:         new.append(char)
   ...:     return ''.join(new)
   ...: 

In [3]: print(strikethrough('strikken'))
s̶t̶r̶i̶k̶k̶e̶n
```

**Comment by geier** [comment]
font support for that method is only so and so:

DejaVu Sans Mono:
![dejavusans](https://cloud.githubusercontent.com/assets/275330/23093329/16cd40de-f5d8-11e6-96ab-6034cbfaa6ed.png)

M+1mn:
![m 1mn](https://cloud.githubusercontent.com/assets/275330/23093332/1e34840e-f5d8-11e6-9d62-cb74f5565ed0.png)

Iosevka:
![iosevka](https://cloud.githubusercontent.com/assets/275330/23093335/2b16f0c6-f5d8-11e6-86c7-4e57d1108142.png)

**Comment by untitaker** [comment]
I think ANSI escape codes are they way to go despite tmux. I wasn't able to find a terminal on Linux that doesn't support strikethrough.

**Comment by geier** [comment]
supporting ANSI strikethrough:
* probably everything based on libvte (tested gnome terminal)
* xterm
* terminology

not supporting:
* urvxt 
* tmux
* screen

**Comment by hobarrera** [comment]
Since, personally, I use `tmux`, I'm really inclined towards a different choice. Strikethrough+symbol might be the best idea since the symbol itself helps differentiate cancelled vs declined.

I'll leave these symbols here for consideration:

⊗⊘ ⊜⊖ ⊝ ⛔ ꚛ ⏹▣ ■ ⯀x × X ✕ ☓ ✖ ✗ ✘

I think the *x mark*s (eg: ✖) make the most sense for declined events, and the first one for cancelled (⊗). This is really subjective though.

(I basically scrolled the entire [unicode symbols table](https://unicode-table.com/en/)).

**Comment by hobarrera** [comment]
While we're talking symbols, we also have the "invited but not accepted" status (`NEEDS-ACTION`?), which  might be worth differentiating from accepted (or "created by me") events.

**Comment by geier** [comment]
Problem with (only using) ANSI strikethrough is, that it gets removed if stdout is not a terminal.

**Comment by geier** [comment]
urwid doesn't seem to support strikethrough either 
https://github.com/urwid/urwid/blob/dc8e8b3fdb0bed9fe6cf8781b4c0b1396b7d406f/urwid/display_common.py

**Comment by geier** [comment]
After tmux, urwid is the most important part to get working now before we can use strikethrough, because I really want to avoid the situation where khal's and ikhal's output look qualitatively different.

**Comment by WhyNotHugo** [comment]
I'd kinda like to bring this issue back to life.

I've been thinking about a configuration option to HIDE declined events altogether (since I've way to many laying around in my work calendar). Do you think that might be acceptable (regardless of any decoration added to them when not-hidden)?

**Comment by geier** [comment]
Sure, go ahead. If you need any help, feel free to contact me.

**Comment by ariep** [comment]
Sorry to add to this old thread, but it appears tmux now supports this out of the box. At least on my system, the `echo -e "\e[9mstrikethrough\e[0m"` test works under tmux without special configuration.

**Comment by geier** [comment]
Can you send some examples of what we are looking for? (Sorry, this issue is ancient, I have forgotten most of the details)

Are we talking about:
 `ATTENDEE;PARTSTAT=DECLINED:mailto:jsmith@example.com`? Do we want to mark events with the user's mail addresses ? (If so, we will need to collect them in the configuration file)

**Comment by geier** [comment]
btw: [urwid now supports strikethrough](https://github.com/urwid/urwid/pull/239), but I'd first go with an icon, as it is easier to implement.

**Comment by madduck** [comment]
Thanks for your response, Chris!

Yes, `ATTENDEE;PARTSTAT=DECLINED:mailto:jsmith@example.com` matching against one of the known email addresses. That is the one use case;

The other use case is an event with `STATUS:CANCELLED`. Also see #587, though I think that DECLINED/CANCELLED are the most important. Just something to keep in mind: there are other statuses as well which might benefit from being exposed.

**Comment by geier** [comment]
Note to self: We do support "status" already, that is, we can print it in the calendar representation. We could 
* print some kind of symbols instead (similar to recurrence symbols)
* make those toggle-able from ikhal

But I'm not sure how STATUS interacts with PARTSTAT. Especially the interface in ikhal is already quite overloaded.

**Comment by geier** [comment]
@madduck  could you have a look if you had something like https://github.com/pimutils/khal/pull/1271 in mind? For now, no strikethrough, just symbols. Also no way to modify, but showing status and partstat as symbols.

**Comment by madduck** [comment]
> @madduck could you have a look if you had something like #1271 in mind? For now, no strikethrough, just symbols. Also no way to modify, but showing status and partstat as symbols.

Sorry, took me a bit longer. It works as you had intended it, and that's great. Thank you for that!

A couple of thoughts:

1. Any reason to use ❔ instead of ?
2. Please don't include a leading space (see https://github.com/pimutils/khal/pull/1271#pullrequestreview-1493002630)
3. Being able to use terminal control sequences to yield italics and strikethrough, similar to how `{calendar_colour}` works, would be awesome.
4. Being able to provide a list of addresses, or even have the addresses read from a separate file or files would make this really scalable.

**Comment by geier** [comment]
Thanks for testing @madduck!

1. no reason really, I'm currently trying out CoPilot for my private coding and it suggested ❔, but will change to ?
2. will fix
3. are you thinking of something you would configure in your khal config? e.g. `[colors] declined = strikethrough`? If so, I would like to handle it differently, I want to expose color theming to the config anyway (because the supplied ones suck for most people) and then create a new attribute for the status and partstat. Those could then be themed. We would need to think about which one should take precedent if both of them are defined.
4. I actually had a list of addresses first, but then I thought we could use this as a stepping stone for getting invitations to work properly and there the (at least easiest) approach would be to have one mail setting per calendar. What would be your use case of several mail addresses per calendar?
